### PR TITLE
Take a stab at rearranging Section 4 et al to reflect the application…

### PR DESCRIPTION
**Pull request by chris-wood**
…/security and security/transport dependencies.

### CODE REVIEW

**Comment by chris-wood** [comment]
@squarooticus can you please take a look before Wednesday?

**Comment by squarooticus** [comment]
Yep, I will review Monday morning.

File: draft-ietf-taps-transport-security.md [hunk 1]
```diff
@@ -676,17 +676,15 @@ in the clear. Everything else is encrypted.
...
 - An unreliable transport protocol such as UDP.
 
-# Common Transport Security Features
+# Security Features and Transport Dependencies
 
 There exists a common set of features shared across the transport protocols surveyed in this document.
-The mandatory features should be provided by any transport security protocol, while the optional features
-are extensions that a subset of the protocols provide. For clarity, we also distinguish between handshake
-and record features.
+Features which have no transport or application dependencies are mandatory. All others are optional, and
```
Comment (squarooticus) [inline draft-ietf-taps-transport-security.md]: I think this statement is confusing. Forward security is a mandatory feature, but certainly depends on some features of the underlying transport or security layer. Is the intended meaning something like

> Mandatory features constitute a baseline of functionality that an application may assume for any TAPS implementation. Optional features by contrast may be available to the application, but must be queried, or the choice of protocols constrained, when their presence or absence impacts the application's use of the transport.

File: draft-ietf-taps-transport-security.md [hunk 2]
```diff
@@ -676,17 +676,15 @@ in the clear. Everything else is encrypted.
...
+Features which have no transport or application dependencies are mandatory. All others are optional, and 
+are commonly part of a subset of protocols described above. We do not distinguish between features provided 
+by different components of a transport security protocol. 
 
 ## Mandatory Features
 
-### Handshake
-
 - Forward-secure segment encryption and authentication: Transit data must be protected with an
 authenticated encryption algorithm.
```
Comment (squarooticus) [inline draft-ietf-taps-transport-security.md]: Something about forward security should be in the description. Maybe add something about "...using keys derived from a forward-secure key exchange".

File: draft-ietf-taps-transport-security.md [hunk 3]
```diff
@@ -696,43 +694,52 @@ many transport security protocols.
...
+- Pre-shared key support: A security protocol must be able to use a pre-shared key established 
+out-of-band or from a prior session to encrypt individual messages, packets, or datagrams.
 
 ## Optional Features
 
-### Handshake
+- Mutual authentication: Transport security protocols must allow each endpoint to authenticate the other if 
+required by the application.
+  - Transport dependency: None.
+  - Application dependency: Mutual authentication required for application support.
```
Comment (squarooticus) [inline draft-ietf-taps-transport-security.md:710]: I'm having a hard time wrapping my head around what is meant by "dependency" in this context. Does "X dependency" mean "X must have some feature" or "X depends on this feature"?

**Review by squarooticus — COMMENTED** [review]
This is improved, but I think we still need to clarify what the features section is trying to achieve. It may be that we need to describe the interfaces between the transport and application bidirectionally, rather than simply as dependencies. For example:

- Mutual authentication in its most general sense has no requirements for the transport beyond the ability to move authentication messages back and forth reliably. In this sense, the retransmission scheme provided by DTLS is actually part of the transport, not part of the security layer.
- Similarly, the transport requires of the security layer only that the authentication messages be expressable in terms that it can transport (typically, a series of datagrams or an octet stream).
- Mutual auth requires of the application that it has access to a cryptographic credential compatible with some scheme supported by the security layer for the purposes of authenticating identity.
- The application requires that the security layer support some mechanism for asserting its identity to the peer and then proving that identity using a credential it has access to, in a way that is cryptographically tied to the session or connection. The first might be a certificate that the peer trusts; the second could then be some signature scheme incorporating a derivation of the master key.

File: draft-ietf-taps-transport-security.md (hunk 3 above)
Comment (chris-wood) [reply draft-ietf-taps-transport-security.md:710]: The former -- for any sort of authentication, we have a hard dependency on the application to provide credentials.

File: draft-ietf-taps-transport-security.md (hunk 2 above)
Comment (chris-wood) [reply draft-ietf-taps-transport-security.md]: Will fix.

File: draft-ietf-taps-transport-security.md (hunk 1 above)
Comment (chris-wood) [reply draft-ietf-taps-transport-security.md]: Yes, that's what I intended. I'll use that text.

File: draft-ietf-taps-transport-security.md [hunk 4]
```diff
@@ -676,63 +676,75 @@ in the clear. Everything else is encrypted.
...
-# Common Transport Security Features
+# Security Features and Transport Dependencies
 
 There exists a common set of features shared across the transport protocols surveyed in this document.
-The mandatory features should be provided by any transport security protocol, while the optional features
-are extensions that a subset of the protocols provide. For clarity, we also distinguish between handshake
-and record features.
+Mandatory features constitute a baseline of functionality that an application may assume for any TAPS
+implementation. Optional features by contrast may be available to an application depending on application
+and transport dependencies. Applications learn of and use optional features by querying for their presence
```
Comment (squarooticus) [inline draft-ietf-taps-transport-security.md]: I'm thinking "Optional features by contrast may vary from implementation to implementation, and so an application cannot simply assume they are available."

**Review by squarooticus — APPROVED** [review]
I think this is fine. We can do further tweaks in subsequent PRs.

**Pull request merged**